### DISCUSSION TITLE
Timeline for multus v4

### DISCUSSION THREAD

**Discussion by maiqueb** [k8snetworkplumbingwg/multus-cni#853] [Q&A] [answered]
As per the subject: when is the release for multus v4 planned ? 

Furthermore, what are the plans for the hotplug feature ? When can I rebase PR #777 and expect it to be timely reviewed ?

**Comment by martinkennelly** [comment]
I see an opportunity to add observability features to multus. Interested in hearing a timeline for v4.

**Comment by s1061123** [comment]
@maiqueb , currently we're working on 4.0 release and about to release. I will finish 4.0 work in July.

However, we will not start to work on the hotplug feature after the 4.0 release unfortunately because we have other prioritized tasks for mutus and work on these first. So we cannot say when we will start to work on that. 

When we start to work on that, we need to think about how to implement (from software architecture point of view, such as implement in multus-daemon or not in multus-daemon). 

Of course, we must follow the interface (API) defined in kubevirt's enhancement doc as we’ve agreed on as a community. However, we’d like to further consider the internal implementation within Multus as the thick plugin aspects in 4.0 comes together, and after some soak in community deployments.

So if you don't mind, we'd like to merge your PR to a new prototype branch and once we start to work on that, we will refer to your prototype.

**Reply by maiqueb** [reply]
> @maiqueb , currently we're working on 4.0 release and about to release. I will finish 4.0 work in July.

Good to know.

> 
> However, we will not start to work on the hotplug feature after the 4.0 release unfortunately because we have other prioritized tasks for mutus and work on these first. So we cannot say when we will start to work on that.
> 
> When we start to work on that, we need to think about how to implement (from software architecture point of view, such as implement in multus-daemon or not in multus-daemon).

In the [August 4th, 2022](https://docs.google.com/document/d/17-vkuoObBFBb6feXNA7tjTaKr-rs8ea0VTnqvlEdCv8/edit#heading=h.quxvqvab41qc) multus maintainer's meeting we have discussed this; we've agreed to open a new API in the multus daemon to invoke a particular delegate, which a 3rd party controller would trigger when it saw the annotation changes.

This would limit - by a lot - the changes in the multus project, while allowing the interested parties to move forward writing this controller.

Later on, once feedback is available, we can re-evaluate if the multus maintainers are interested in implementing this feature (or not).

**Reply by maiqueb** [reply]
For completeness; the aforementioned controller is available in this [project](https://github.com/k8snetworkplumbingwg/multus-dynamic-networks-controller).

**Comment by Alan01252** [comment]
Are there any updates please on when there could be an official release of 4.0?

**Reply by maiqueb** [reply]
This is being held back by https://github.com/k8snetworkplumbingwg/multus-cni/pull/1006; I've spoken w/ @dougbtv / @s1061123  and I'll take over this PR the next few weeks.

You can track the following issue for progress: #1050 .

**Comment by maiqueb** [comment, marked as answer]
[v4](https://github.com/k8snetworkplumbingwg/multus-cni/releases/tag/v4.0.0) is out there!
